Replace Buggy DT/Camera with SharkPool/Camera

SharkPool-SP opened the pull request:
Resolves #930
Resolves #1839 

Remake of DTs buggy camera extension with inspiration from PenguinMods camera extension.

This fixes all known issues with DT's camera. Interpolation support is still yet to be finished but GarboMuffin needs to add a api to allow inserting interpolation data

:)

![image](https://github.com/user-attachments/assets/605ca461-032f-4d58-823b-0b700e761eec)

SharkPool-SP commented:
!format

SharkPool-SP commented:
!format

hammouda101010 commented:
can't wait for someone to merge this

CST1229 commented on extensions/SharkPool/Camera.js:
```diff
@@ -0,0 +1,795 @@
+  // custom gui
+  function openModal(titleName, func) {
+    // in a Button Context, ScratchBlocks always exists
+    ScratchBlocks.Variables.createVariable(
+      ScratchBlocks.mainWorkspace,
+      null,
+      "broadcast_msg"
+    );
```
probably just nitpicking but i'm not sure if this is the best way to make a scratch-blocks modal (as opposed to, for example, using `ScratchBlocks.prompt("title", "default value", (value) => console.log(value), "title", "broadcast_msg")` directly)

SharkPool-SP replied:
tbh didnt even know that existed till today

SharkPool-SP commented:
https://github.com/TurboWarp/extensions/issues/1839

CubesterYT commented on extensions/SharkPool/Camera.js:
```diff
@@ -0,0 +1,788 @@
+
+  const menuIconURI =
+    "data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI0MSIgaGVpZ2h0PSI0MSIgdmlld0JveD0iMCAwIDQxIDQxIj48ZyBzdHJva2Utd2lkdGg9IjAiIHN0cm9rZS1taXRlcmxpbWl0PSIxMCI+PHBhdGggZD0iTTAgMjAuNUMwIDkuMTc4IDkuMTc4IDAgMjAuNSAwUzQxIDkuMTc4IDQxIDIwLjUgMzEuODIyIDQxIDIwLjUgNDEgMCAzMS44MjIgMCAyMC41IiBmaWxsPSIjMjg1MWM5Ii8+PHBhdGggZD0iTTIuMzc4IDIwLjVjMC0xMC4wMDkgOC4xMTMtMTguMTIy ...
+  const cameraIcon =
```
Since the Camera icon is used only in a couple blocks, just don't use it at all. It looks nice with no icon.

CubesterYT replied:
Or apply the icon to all blocks

CubesterYT commented on extensions/SharkPool/Camera.js:
```diff
@@ -0,0 +1,788 @@
+
+      if (runtime.ext_videoSensing)
+        objectNames.push({
+          text: Scratch.translate("Video Layer"),
```
```diff
- text: Scratch.translate("Video Layer"),
+ text: Scratch.translate("video layer"),
```

CubesterYT commented on extensions/SharkPool/Camera.js:
```diff
@@ -0,0 +1,788 @@
+        });
+      if (runtime.ext_pen)
+        objectNames.push({
+          text: Scratch.translate("Pen Layer"),
```
```diff
- text: Scratch.translate("Pen Layer"),
+ text: Scratch.translate("pen layer"),
```

CubesterYT reviewed (changes requested):
First round

CubesterYT commented on extensions/SharkPool/Camera.js:
```diff
@@ -0,0 +1,788 @@
+          text: Scratch.translate("all objects"),
+          value: "_all_",
+        });
+      objectNames.push({ text: Scratch.translate("Stage"), value: "_stage_" });
```
```diff
- text: Scratch.translate("Stage"),
+ text: Scratch.translate("stage"),
```

SharkPool-SP commented:
!format

SharkPool-SP commented:
!format

SharkPool-SP commented:
!format

gomigg07 commented:
Just throwing this out there, I think adding a `set scroll factor` block would work wonders. For example, if the scroll factor in a sprite or clone were set to `0.5` (or `50%` going off of Scratch measurements), the sprite would be offset by half. Moving the camera 100px to the right displays the sprite only moving 50px to the left. I think it would be useful for parallax effects like scrolling backgrounds.

SharkPool-SP commented:
> Just throwing this out there, I think adding a `set scroll factor` block would work wonders. For example, if the scroll factor in a sprite or clone were set to `0.5` (or `50%` going off of Scratch measurements), the sprite would be offset by half the value of the camera's position. Moving the camera 100px to the right displays the sprite only moving 50px to the left. I think it would be useful for parallax effects like scrolling backgrounds.

Why not just make a separate camera for that sprite/clone

gomigg07 commented:
> Why not just make a separate camera for that sprite/clone

Yeah, that's fair.

fg133266 commented:
https://github.com/user-attachments/assets/38d4f720-11c7-442a-997e-56f34ce0a929

I'm trying to bind the clone arrow in the video to some specified Camera, but its xy doesn't seem to be accurate, it only works when I try to update the Camera, am I doing something wrong? ;-;

SharkPool-SP commented:
> 
> 
> https://github.com/user-attachments/assets/38d4f720-11c7-442a-997e-56f34ce0a929
> 
> 
> 
> I'm trying to bind the clone arrow in the video to some specified Camera, but its xy doesn't seem to be accurate, it only works when I try to update the Camera, am I doing something wrong? ;-;

Try pausing the arrow movement and  setting the camera position to random amounts. Also in the video it's not clear whether the camera x started at 100

fg133266 commented:
> > QQ202521-173839.mp4
> >     
> >   
> > 
> >   
> > 
> >   
> > 
> > 
> > 我正在尝试将视频中的克隆箭头绑定到某个指定的 Camera，但它的 xy 似乎不准确，它仅在我尝试更新 Camera 时有效，是不是我做错了什么？ ;-;
> 
> 尝试暂停箭头移动并将摄像机位置设置为随机量。此外，在视频中，尚不清楚摄像头 x 是否从 100 开始

![Snipaste_2025-02-02_06-48-07](https://github.com/user-attachments/assets/6c8a3372-1b51-4438-962d-001c63ffd843)
![Snipaste_2025-02-02_06-56-59](https://github.com/user-attachments/assets/60271933-fe5c-4171-a00a-ff913f2f3298)

Yes, the x of the camera I set it to 100 at the beginning, and then I added the blocks in the diagram and removed the blocks related to the X movement, and strangely the arrow was at 0 instead of 100

SharkPool-SP commented:
> > > QQ202521-173839.mp4
> 
> > >     
> 
> > >   
> 
> > > 
> 
> > >   
> 
> > > 
> 
> > >   
> 
> > > 
> 
> > > 
> 
> > > 我正在尝试将视频中的克隆箭头绑定到某个指定的 Camera，但它的 xy 似乎不准确，它仅在我尝试更新 Camera 时有效，是不是我做错了什么？ ;-;
> 
> > 
> 
> > 尝试暂停箭头移动并将摄像机位置设置为随机量。此外，在视频中，尚不清楚摄像头 x 是否从 100 开始
> 
> 
> 
> ![Snipaste_2025-02-02_06-48-07](https://github.com/user-attachments/assets/6c8a3372-1b51-4438-962d-001c63ffd843)
> 
> ![Snipaste_2025-02-02_06-56-59](https://github.com/user-attachments/assets/60271933-fe5c-4171-a00a-ff913f2f3298)
> 
> 
> 
> Yes, the x of the camera I set it to 100 at the beginning, and then I added the blocks in the diagram and removed the blocks related to the X movement, and strangely the arrow was at 0 instead of 100

I don't understand what you're talking about

fg133266 commented:
> > > > QQ202521-173839.mp4
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > 我正在尝试将视频中的克隆箭头绑定到某个指定的 Camera，但它的 xy 似乎不准确，它仅在我尝试更新 Camera 时有效，是不是我做错了什么？ ;-;
> > 
> > 
> > > 
> > 
> > 
> > > 尝试暂停箭头移动并将摄像机位置设置为随机量。此外，在视频中，尚不清楚摄像头 x 是否从 100 开始
> > 
> > 
> > ![Snipaste_2025-02-02_06-48-07](https://private-user-images.githubusercontent.com/143243975/408830829-6c8a3372-1b51-4438-962d-001c63ffd843.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg0NjIyOTcsIm5iZiI6MTczODQ2MTk5NywicGF0aCI6Ii8xNDMyNDM5NzUvNDA4ODMwODI5LTZjOGEzMzcyLTFiNTEtNDQzOC05NjJkLTAwMWM2M2ZmZDg0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwMlQwMjA2MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDUxYzk0Zjc2YThlNzM0YTYyNjAzMWMzNjA3NTVkZTllM2QwZGJjZDBjNjAwYjI3YTE5M2NmM2M4ZjhiNzk2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0R8Bw-PXM8rAeMAWZJTNZzQ3UKFHdXb8NrpS3ybnqAM)
> > ![Snipaste_2025-02-02_06-56-59](https://private-user-images.githubusercontent.com/143243975/408831328-60271933-fe5c-4171-a00a-ff913f2f3298.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg0NjIyOTcsIm5iZiI6MTczODQ2MTk5NywicGF0aCI6Ii8xNDMyNDM5NzUvNDA4ODMxMzI4LTYwMjcxOTMzLWZlNWMtNDE3MS1hMDBhLWZmOTEzZjJmMzI5OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwMlQwMjA2MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYzJiNDM4MWZiYjNkMDQ0MjY2NGVjMzU4M2YwMTJmZTQzYjQ3MTE2ZWMwY2E0M2M4ZDM3ZmYwYTFjNTYyNDllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Sno2-YZ9l4FzajHxATj4wUR1Yhnu6osKd9ShaYzNbMY)
> > 是的，相机的 x 我一开始就把它设置为 100，然后我在图表中添加了块并删除了与 X 运动相关的块，奇怪的是箭头在 0 而不是 100 处
> 
> 我不明白你在说什么

Oh no

fg133266 commented:
![Snipaste_2025-02-02_10-51-48](https://github.com/user-attachments/assets/ca67e1dd-ffc5-4295-be42-18ed3a9cfccf)
我认为应该是这样导致出现了问题，对不起;-;

fg133266 commented:
![Snipaste_2025-02-02_10-51-48](https://github.com/user-attachments/assets/e425235a-5217-4718-b8ac-81971ad00d68)
I think this should have caused the problem. I'm sorry;-;

fg133266 commented:
![Snipaste_2025-02-02_11-00-26](https://github.com/user-attachments/assets/8443ef73-0696-4a97-9494-ab67a01bbe38)
I think this should have caused the problem. I'm sorry;-;

AD1340 commented:
Please...

SharkPool-SP commented:
> ![Snipaste_2025-02-02_11-00-26](https://private-user-images.githubusercontent.com/143243975/408839711-8443ef73-0696-4a97-9494-ab67a01bbe38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg1NjcxOTMsIm5iZiI6MTczODU2Njg5MywicGF0aCI6Ii8xNDMyNDM5NzUvNDA4ODM5NzExLTg0NDNlZjczLTA2OTYtNGE5Ny05NDk0LWFiNjdhMDFiYmUzOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwM1QwNzE0NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZDJjNGMxYjc0ZDI0NDljMGE3ZTcyZGY3ODRmNWY0ZjI5NWZlZDBhMjM3ZWI2NjIxY2U2MzU4ZDQwOTg5MDJlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7l2yX_w8kT2nZB63QkzWqe6lB8jlRYDOyhh5dR_TM8g) I think this should have caused the problem. I'm sorry;-;

its a bug with my reset code. ill fix it in a bit

CubesterYT reviewed (approved):
Everything seems good, but don't take my word for it

I could very well have missed something.

CST1229 commented on extensions/SharkPool/Camera.js:
```diff
@@ -0,0 +1,795 @@
+  }
+
+  function bindDrawable(drawable, camera) {
+    const camSystem = drawable[cameraSymbol];
+    if (camSystem.name === camera) return;
```
(this one no longer counts)

SharkPool-SP commented:
!format

AD1340 commented:
Yesyesyesyes

CubesterYT commented:
I think it's time, we had our first new extension after a long while.
